### PR TITLE
fix #14522, litterals > int64.high are now uint64

### LINKED ISSUE #14522: Valid uint64 literal says -72057594037927936 can't be converted to uint64.
```nim
echo 0xFF000000_00000000.uint64
```
```
Error: -72057594037927936 can't be converted to uint64
```
https://play.nim-lang.org/#ix=2nY6

I think its treating all hex literals as signed int64, so it can't accept one that has a negative bit set. Some where in some conversions this happens?

My compiler:
```
Nim Compiler Version 1.2.0 [Windows: amd64]
Compiled at 2020-04-03
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 7e83adff84be5d0c401a213eccb61e321a3fb1ff
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by timotheecour**
fix #14522
same rationale as https://github.com/nim-lang/Nim/blob/de27910ab72dca681188953ae963802520271e43/compiler/lexer.nim#L603

## mildly related links
* https://github.com/nim-lang/Nim/issues/14529
* https://github.com/nim-lang/RFCs/issues/228
* https://github.com/timotheecour/Nim/issues/125

### CODE REVIEW

File: compiler/lexer.nim [hunk 1]
```diff
@@ -572,18 +580,27 @@ proc getNumber(L: var TLexer, result: var TToken) =
...
+        if uint64Compatible and not ok:
+          try:
+            var iNumber2: uint64
+            len = parseBiggestUInt(result.literal, iNumber2)
+            iNumber = cast[int64](iNumber2)
+            ok = true
+            result.tokType = tkUInt64Lit
+          except ValueError: discard
+        if not ok:
           raise newException(OverflowDefect, "number out of range: " & $result.literal)
```
Comment (alaviss) [inline compiler/lexer.nim:602]: Alternative implementation:

```nim
        try:
          try:
            len = parseBiggestInt(result.literal, iNumber)
          except ValueError:
            if uint64Compatible:
              var uiNumber: uint64
              len = parseBiggestUint(result.literal, uiNumber)
              iNumber = cast[int64](uiNumber)
              result.tokType = tkUInt64Lit
            else:
              raise
        except ValueError:
          raise newException(OverflowDefect, "number out of range: " & $result.literal)
```

This way there's no need to introduce an `ok` variable. But of course I'd say this one should be up to taste.

IMO you should change `iNumber2` into `uiNumber` though.

File: lib/pure/parseutils.nim [hunk 2]
```diff
@@ -393,7 +393,7 @@ proc captureBetween*(s: string, first: char, second = '\0', start = 0): string =
   result = ""
   discard s.parseUntil(result, if second == '\0': first else: second, i)
 
-proc integerOutOfRangeDefect() {.noinline.} =
+proc integerOutOfRangeError() {.noinline.} =
```
Comment (alaviss) [inline lib/pure/parseutils.nim:396]: Can you please not conflate bugfixes with refactoring?

File: testament/categories.nim [hunk 3]
```diff
@@ -611,7 +611,7 @@ proc runJoinedTest(r: var TResults, cat: Category, testsDir: string) =
   #[
   TODO(minor):
   get from Nim cmd
-  put outputGotten.txt, outputGotten.txt, megatest.nim there too
+  put outputExceptedFile, outputGottenFile, megatest.nim there too
```
Comment (alaviss) [inline testament/categories.nim:614]: I'd appreciate it more if unrelated changes can be moved into a separated PR.

File: testament/lib/stdtest/testutils.nim [hunk 4]
```diff
@@ -23,3 +23,14 @@ template flakyAssert*(cond: untyped, msg = "", notifySuccess = true) =
...
+import macros
+
+template doAssertParserRaises*(exception: typedesc, s: static string) =
+  # xxx not sure if there's a simpler way to pass a type (not as NimNode) to a macro
+  # but this works. Note: generic macros would solve this and other issues
+  # cleanly.
+  block:
+    macro doAssertParserRaises2(s2: static string): void =
+      doAssertRaises(exception): discard parseExpr(s2)
+    doAssertParserRaises2(s)
```
Comment (alaviss) [inline testament/lib/stdtest/testutils.nim:36]: Better naming and clean up.

```suggestion
    macro doAssertParserRaisesImpl() =
      doAssertRaises(exception): discard parseExpr(s)
    doAssertParserRaisesImpl()
```

You may also be able to remove `block:` by adding `{.gensym.}` to the macro.

**Comment by alaviss** [comment]
I'm not sure if we want this kind of "promotion" to `uint`. This open quite a can of worms, since `uint` semantics is vastly different from `int`.

Instead we might want to reject and produce a `Hint` telling the developer to annotate their type as `uint` via the `u*` suffixies.

We may also want to defer auto conversion to the biggest `int` type and see if the programmer decided to convert the `int` into an another type (like `uint64`), then generate a literal of that type directly. Given that we now have `int128` in the compiler, we can make `int literal` itself `int128` then resolve the native type upon usage.

File: lib/pure/parseutils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/parseutils.nim:396]: fine, i'll move those refactorings to a separate PR

**Comment by Araq** [comment]
Somewhat related: I wonder if "length-specific hex literals" are intuitive:

0xff # ok, uint8
0x00ff # ok, uint16
0x00_00_ff # invalid
0x00_00_ff_00 # uint32

etc

**Comment by timotheecour** [comment]
> Somewhat related: I wonder if "length-specific hex literals" are intuitive:

yes, but that's intimately related to whether hex (or binary) litterals are typed are unsigned, which is the most intuitive behavior (and can do some research on how other languages treat those; if you want signed you'll have to use `0xffi8` otherwise `0xff` would be unsigned)

`0x00_00_ff # invalid`

yes! this has no reason for existance. This would at least fix https://github.com/nim-lang/Nim/issues/14529

so we are getting somewhere now, the length encoding you're suggesting solves the problem you raised in https://github.com/nim-lang/Nim/issues/14529#issuecomment-636653280

```
0xf # ERROR
0xff # ok, uint8 (instead of int)
0xffi8 # ok, int8
...
0x00_00_ff_00 # uint32 (instead of int)
0x00_00_ff_00_00_00_ff_00 # uint64 (instead of int64 <= NOTICE the pre-existing inconsistency here)

# explicit suffix is ok:
0xffi32 # ok, int32 instead of uint8)
0xffu32 # ok, uint32 (instead of uint8)
0xfu32 # still ERROR
```

ditto with binary

## analog to `\x HHH` became illegal
analog to how (2 y ago i forgot) we made it an error to have `\x HHH`
```
\x HH | character with hex value HH; exactly two hex digits are allowed
-- | --
\u HHHH | unicode codepoint with hex value HHHH; exactly four hex digits are allowed

```

**Comment by Clyybber** [comment]
I think this is better solved with https://github.com/nim-lang/RFCs/issues/216 / https://github.com/nim-lang/RFCs/issues/228 

Theres still the question of a default though.

**Comment by timotheecour** [comment]
> I think this is better solved with

araq's "length-specific hex literals" should be 100% compatible with at least this RFC https://github.com/nim-lang/RFCs/issues/228
here the question is for when you dont have a suffix
